File: android/app/src/main/java/com/apoorvdarshan/calorietracker/models/AddMenuConfig.kt
package com.apoorvdarshan.calorietracker.models

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import com.apoorvdarshan.calorietracker.R
import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class AddMenuGroupConfig(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val methods: List<String>
)

@Serializable
data class AddMenuConfig(
    val version: Int = CURRENT_VERSION,
    val groups: List<AddMenuGroupConfig> = emptyList(),
    val flatMethods: List<String> = emptyList()
) {
    val usesFlatLayout: Boolean get() = groups.isEmpty()

    fun sanitized(): AddMenuConfig {
        val seen = mutableSetOf<FoodLogMethod>()

        fun filterMethods(raw: List<String>): List<String> {
            val result = mutableListOf<String>()
            for (key in raw) {
                val method = FoodLogMethod.fromStorage(key) ?: continue
                if (method !in seen) {
                    seen += method
                    result += method.storageKey
                }
            }
            return result
        }

        val trimmedGroups = groups.take(3).mapNotNull { group ->
            val methods = filterMethods(group.methods)
            if (methods.isEmpty()) return@mapNotNull null
            AddMenuGroupConfig(
                id = group.id.ifBlank { UUID.randomUUID().toString() },
                name = group.name.trim().ifBlank { DEFAULT_GROUP_FALLBACK_NAME },
                methods = methods
            )
        }

        return if (trimmedGroups.isEmpty()) {
            copy(version = CURRENT_VERSION, groups = emptyList(), flatMethods = filterMethods(flatMethods))
        } else {
            copy(version = CURRENT_VERSION, groups = trimmedGroups, flatMethods = emptyList())
        }
    }

    fun resolvedGroups(): List<ResolvedAddMenuGroup> =
        groups.mapNotNull { group ->
            val methods = group.methods.mapNotNull(FoodLogMethod::fromStorage)
            if (methods.isEmpty()) return@mapNotNull null
            ResolvedAddMenuGroup(
                id = group.id,
                name = group.name,
                methods = methods
            )
        }

    fun resolvedFlatMethods(): List<FoodLogMethod> =
        flatMethods.mapNotNull(FoodLogMethod::fromStorage)

    /**
     * Puts a hidden method back on the menu.
     *
     * - [groupIndex] set: add to that group (and drop the key from every other group).
     * - Flat layout: append to [flatMethods].
     * - Grouped layout with no index: last group, or the first group that already
     *   has methods if the last one is empty.
     */
    fun withRestoredMethod(
        method: FoodLogMethod,
        groupIndex: Int? = null
    ): AddMenuConfig {
        val key = method.storageKey
        val targetIndex = when {
            groupIndex != null && groupIndex in groups.indices -> groupIndex
            usesFlatLayout -> null
            groups.isEmpty() -> null
            groups.last().methods.isNotEmpty() -> groups.lastIndex
            else -> groups.indexOfFirst { it.methods.isNotEmpty() }.takeIf { it >= 0 }
                ?: groups.lastIndex
        }
        if (targetIndex == null) {
            return copy(flatMethods = flatMethods + key).sanitized()
        }
        val nextGroups = groups.mapIndexed { idx, group ->
            if (idx == targetIndex) {
                group.copy(methods = group.methods + key)
            } else {
                group.copy(methods = group.methods.filterNot { it == key })
            }
        }
        return copy(
            groups = nextGroups,
            flatMethods = flatMethods.filterNot { it == key }
        ).sanitized()
    }

    companion object {
        const val CURRENT_VERSION = 1
        const val STORAGE_KEY = "addMenu.config"

        /** Stable identifiers for default group names — resolved to localized strings at display time. */
        const val DEFAULT_GROUP_PHOTO_SCAN = "__default_photo_scan__"
        const val DEFAULT_GROUP_DESCRIBE_MEAL = "__default_describe_meal__"
        const val DEFAULT_GROUP_REUSE_MEAL = "__default_reuse_meal__"
        const val DEFAULT_GROUP_FALLBACK_NAME = "__default_group__"

        /** Matches the pre-customization Android Home + food menu. */
        val Default = AddMenuConfig(
            groups = listOf(
                AddMenuGroupConfig(
                    name = DEFAULT_GROUP_PHOTO_SCAN,
                    methods = listOf(
                        FoodLogMethod.CAMERA.storageKey,
                        FoodLogMethod.PHOTOS.storageKey,
                        FoodLogMethod.BARCODE.storageKey
                    )
                ),
                AddMenuGroupConfig(
                    name = DEFAULT_GROUP_DESCRIBE_MEAL,
                    methods = listOf(
                        FoodLogMethod.TEXT.storageKey,
                        FoodLogMethod.VOICE.storageKey,
                        FoodLogMethod.MANUAL.storageKey
                    )
                ),
                AddMenuGroupConfig(
                    name = DEFAULT_GROUP_REUSE_MEAL,
                    methods = listOf(
                        FoodLogMethod.RECENT.storageKey,
                        FoodLogMethod.FREQUENT.storageKey,
                        FoodLogMethod.FAVORITES.storageKey,
                        FoodLogMethod.COPY_FROM_DAY.storageKey
                    )
                )
            )
        )

        fun decode(raw: String?): AddMenuConfig {
            if (raw.isNullOrBlank()) return Default
            return runCatching {
                val decoded = json.decodeFromString<AddMenuConfig>(raw)
                val sanitized = decoded.sanitized()
                if (sanitized.groups.isEmpty() && sanitized.flatMethods.isEmpty()) {
                    val hadConfiguredContent = decoded.groups.any { it.methods.isNotEmpty() } ||
                        decoded.flatMethods.isNotEmpty()
                    if (hadConfiguredContent) Default else sanitized
                } else {
                    sanitized
                }
            }.getOrDefault(Default)
        }

        fun encode(config: AddMenuConfig): String =
            json.encodeToString(serializer(), config.sanitized())

        /**
         * Settings startup reads the menu, then later replaces the whole screen state.
         * If the user already edited during that gap, keep the menu on screen.
         */
        fun afterSettingsSnapshot(
            alreadyLoaded: Boolean,
            onScreen: AddMenuConfig,
            diskRead: AddMenuConfig,
        ): AddMenuConfig = if (alreadyLoaded) onScreen else diskRead

        private val json = kotlinx.serialization.json.Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
        }
    }
}

data class ResolvedAddMenuGroup(
    val id: String,
    val name: String,
    val methods: List<FoodLogMethod>
)

internal fun defaultGroupNameRes(storedName: String): Int? = when (storedName) {
    AddMenuConfig.DEFAULT_GROUP_PHOTO_SCAN -> R.string.home_menu_photo_scan
    AddMenuConfig.DEFAULT_GROUP_DESCRIBE_MEAL -> R.string.home_menu_describe_meal
    AddMenuConfig.DEFAULT_GROUP_REUSE_MEAL -> R.string.home_menu_reuse_meal
    AddMenuConfig.DEFAULT_GROUP_FALLBACK_NAME -> R.string.settings_add_menu_new_group
    else -> null
}

@Composable
fun AddMenuGroupConfig.displayName(): String {
    defaultGroupNameRes(name)?.let { return stringResource(it) }
    return name
}

@Composable
fun ResolvedAddMenuGroup.displayName(): String {
    defaultGroupNameRes(name)?.let { return stringResource(it) }
    return name
}

File: ios/calorietracker/Models/AddMenuConfig.swift
import Foundation

struct AddMenuGroupConfig: Codable, Equatable, Identifiable, Sendable {
    var id: UUID
    var name: String
    var methods: [FoodLogMethod]

    init(id: UUID = UUID(), name: String, methods: [FoodLogMethod]) {
        self.id = id
        self.name = name
        self.methods = methods
    }
}

struct AddMenuConfig: Codable, Equatable, Sendable {
    static let currentVersion = 1
    static let storageKey = "addMenu.config"

    var version: Int
    var groups: [AddMenuGroupConfig]
    /// Used when `groups` is empty — flat list under + with no nesting.
    var flatMethods: [FoodLogMethod]

    init(version: Int = currentVersion, groups: [AddMenuGroupConfig], flatMethods: [FoodLogMethod] = []) {
        self.version = version
        self.groups = groups
        self.flatMethods = flatMethods
    }

    /// Matches the pre-customization iOS Home + food menu.
    static let iOSDefault = AddMenuConfig(
        groups: [
            AddMenuGroupConfig(
                name: String(localized: "Reuse Meal", comment: "Default add menu group"),
                methods: [.copyFromDay, .favorites, .frequent, .recent]
            ),
            AddMenuGroupConfig(
                name: String(localized: "Describe Meal", comment: "Default add menu group"),
                methods: [.manual, .siriPhrases, .voice, .text]
            ),
            AddMenuGroupConfig(
                name: String(localized: "Photo & Scan", comment: "Default add menu group"),
                methods: [.barcode, .photos, .camera]
            ),
        ]
    )

    var usesFlatLayout: Bool { groups.isEmpty }

    var visibleMethods: [FoodLogMethod] {
        if usesFlatLayout {
            return flatMethods
        }
        return groups.flatMap(\.methods)
    }

    func sanitized() -> AddMenuConfig {
        let allowed = Set(FoodLogMethod.addMenuCases)
        var seen = Set<FoodLogMethod>()

        func filterMethods(_ methods: [FoodLogMethod]) -> [FoodLogMethod] {
            methods.compactMap { method in
                guard allowed.contains(method), !seen.contains(method) else { return nil }
                seen.insert(method)
                return method
            }
        }

        let trimmedGroups = groups.prefix(3).map { group in
            AddMenuGroupConfig(
                id: group.id,
                name: group.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                    ? String(localized: "Group", comment: "Fallback add menu group name")
                    : group.name.trimmingCharacters(in: .whitespacesAndNewlines),
                methods: filterMethods(group.methods)
            )
        }

        if trimmedGroups.isEmpty {
            return AddMenuConfig(
                version: Self.currentVersion,
                groups: [],
                flatMethods: filterMethods(flatMethods)
            )
        }

        return AddMenuConfig(
            version: Self.currentVersion,
            groups: trimmedGroups,
            flatMethods: []
        )
    }
}

enum AddMenuSettings {
    static func load(store: UserDefaults = .standard) -> AddMenuConfig {
        guard let data = store.data(forKey: AddMenuConfig.storageKey),
              let decoded = try? JSONDecoder().decode(AddMenuConfig.self, from: data)
        else {
            return .iOSDefault
        }
        let sanitized = decoded.sanitized()
        if sanitized.groups.isEmpty, sanitized.flatMethods.isEmpty, !decoded.groups.isEmpty {
            return .iOSDefault
        }
        return sanitized
    }

    static func save(_ config: AddMenuConfig, store: UserDefaults = .standard) {
        let sanitized = config.sanitized()
        guard let data = try? JSONEncoder().encode(sanitized) else { return }
        store.set(data, forKey: AddMenuConfig.storageKey)
    }

    static func reset(store: UserDefaults = .standard) {
        store.removeObject(forKey: AddMenuConfig.storageKey)
    }
}
